File: HYKA/generate_apple_jwt.py
#!/usr/bin/env python3
"""
Generate Apple Sign In JWT Client Secret

Usage:
    python3 generate_apple_jwt.py <TEAM_ID> <SERVICE_ID> [KEY_FILE]

Example:
    python3 generate_apple_jwt.py ABC123XYZ com.hyka.signin
"""

import sys
import json
import time
from datetime import datetime, timedelta
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.backends import default_backend
import jwt

# Default values
DEFAULT_KEY_FILE = '/Users/ricardoda-silva/Downloads/AuthKey_GTT69XWZG8.p8'
KEY_ID = 'GTT69XWZG8'

def generate_jwt(team_id, service_id, key_file=DEFAULT_KEY_FILE):
    """Generate JWT client secret for Apple Sign In"""
    
    # Read private key
    try:
        with open(key_file, 'r') as f:
            private_key_pem = f.read()
    except FileNotFoundError:
        print(f'❌ Key file not found: {key_file}')
        sys.exit(1)
    
    # Load private key
    try:
        private_key = serialization.load_pem_private_key(
            private_key_pem.encode(),
            password=None,
            backend=default_backend()
        )
    except Exception as e:
        print(f'❌ Error loading private key: {e}')
        sys.exit(1)
    
    # Create JWT header
    header = {
        'alg': 'ES256',
        'kid': KEY_ID
    }
    
    # Create JWT payload
    now = int(time.time())
    exp = now + (6 * 30 * 24 * 60 * 60)  # 6 months from now
    
    payload = {
        'iss': team_id,
        'iat': now,
        'exp': exp,
        'aud': 'https://appleid.apple.com',
        'sub': service_id
    }
    
    print('🔑 Generating Apple Sign In JWT Client Secret...')
    print('')
    print('Configuration:')
    print(f'   Team ID:    {team_id}')
    print(f'   Service ID: {service_id}')
    print(f'   Key ID:     {KEY_ID}')
    print(f'   Key File:   {key_file}')
    print('')
    
    # Generate JWT
    try:
        token = jwt.encode(
            payload,
            private_key,
            algorithm='ES256',
            headers=header
        )
        
        print('✅ JWT Generated Successfully!')
        print('')
        print('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━')
        print('APPLE SIGN IN CLIENT SECRET (JWT):')
        print('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━')
        print('')
        print(token)
        print('')
        print('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━')
        print('')
        print('📋 JWT Details:')
        print(f'   Issued At:  {datetime.fromtimestamp(now).isoformat()}')
        print(f'   Expires:    {datetime.fromtimestamp(exp).isoformat()}')
        print(f'   Valid for:  6 months')
        print('')
        print('💡 Use this JWT as the "Client Secret" in Supabase Dashboard → Authentication → Providers → Apple')
        print('')
        
        return token
        
    except Exception as e:
        print(f'❌ Error generating JWT: {e}')
        print('')
        print('Make sure:')
        print('  1. The .p8 file is valid')
        print('  2. The Team ID is correct')
        print('  3. The Service ID is correct')
        print('  4. You have PyJWT and cryptography installed:')
        print('     pip3 install PyJWT cryptography')
        sys.exit(1)

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print('❌ Missing required arguments')
        print('')
        print('Usage: python3 generate_apple_jwt.py <TEAM_ID> <SERVICE_ID> [KEY_FILE]')
        print('')
        print('Example:')
        print('  python3 generate_apple_jwt.py ABC123XYZ com.hyka.signin')
        print('')
        print('Arguments:')
        print('  TEAM_ID    - Your Apple Team ID (found in Apple Developer Portal, top right)')
        print('  SERVICE_ID - Your Service ID for Sign in with Apple')
        print('  KEY_FILE   - Path to .p8 file (optional, defaults to AuthKey_GTT69XWZG8.p8)')
        print('')
        print('Install dependencies:')
        print('  pip3 install PyJWT cryptography')
        sys.exit(1)
    
    team_id = sys.argv[1]
    service_id = sys.argv[2]
    key_file = sys.argv[3] if len(sys.argv) > 3 else DEFAULT_KEY_FILE
    
    generate_jwt(team_id, service_id, key_file)

File: HYKA/generate_apple_jwt.js
#!/usr/bin/env node

/**
 * Generate Apple Sign In JWT Client Secret
 * 
 * Usage:
 *   node generate_apple_jwt.js <TEAM_ID> <SERVICE_ID>
 * 
 * Example:
 *   node generate_apple_jwt.js ABC123XYZ com.hyka.signin
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Get arguments
const teamId = process.argv[2];
const serviceId = process.argv[3];
const keyFile = process.argv[4] || '/Users/ricardoda-silva/Downloads/AuthKey_GTT69XWZG8.p8';
const keyId = 'GTT69XWZG8';

if (!teamId || !serviceId) {
  console.error('❌ Missing required arguments');
  console.error('');
  console.error('Usage: node generate_apple_jwt.js <TEAM_ID> <SERVICE_ID> [KEY_FILE]');
  console.error('');
  console.error('Example:');
  console.error('  node generate_apple_jwt.js ABC123XYZ com.hyka.signin');
  console.error('');
  console.error('Arguments:');
  console.error('  TEAM_ID    - Your Apple Team ID (found in Apple Developer Portal, top right)');
  console.error('  SERVICE_ID - Your Service ID for Sign in with Apple');
  console.error('  KEY_FILE   - Path to .p8 file (optional, defaults to AuthKey_GTT69XWZG8.p8)');
  process.exit(1);
}

// Read private key
if (!fs.existsSync(keyFile)) {
  console.error(`❌ Key file not found: ${keyFile}`);
  process.exit(1);
}

const privateKeyContent = fs.readFileSync(keyFile, 'utf8');

// Create JWT header
const header = {
  alg: 'ES256',
  kid: keyId
};

// Create JWT payload
const now = Math.floor(Date.now() / 1000);
const exp = now + (6 * 30 * 24 * 60 * 60); // 6 months from now

const payload = {
  iss: teamId,
  iat: now,
  exp: exp,
  aud: 'https://appleid.apple.com',
  sub: serviceId
};

console.log('🔑 Generating Apple Sign In JWT Client Secret...');
console.log('');
console.log('Configuration:');
console.log(`   Team ID:    ${teamId}`);
console.log(`   Service ID: ${serviceId}`);
console.log(`   Key ID:     ${keyId}`);
console.log(`   Key File:   ${keyFile}`);
console.log('');

// Sign JWT
try {
  const headerBase64 = Buffer.from(JSON.stringify(header)).toString('base64url');
  const payloadBase64 = Buffer.from(JSON.stringify(payload)).toString('base64url');
  const unsignedToken = `${headerBase64}.${payloadBase64}`;

  // Sign with ES256
  const sign = crypto.createSign('SHA256');
  sign.update(unsignedToken);
  sign.end();
  
  const signature = sign.sign({
    key: privateKeyContent,
    dsaEncoding: 'ieee-p1363'
  }, 'base64url');

  const jwt = `${unsignedToken}.${signature}`;

  console.log('✅ JWT Generated Successfully!');
  console.log('');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('APPLE SIGN IN CLIENT SECRET (JWT):');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('');
  console.log(jwt);
  console.log('');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('');
  console.log('📋 JWT Details:');
  console.log(`   Issued At:  ${new Date(now * 1000).toISOString()}`);
  console.log(`   Expires:    ${new Date(exp * 1000).toISOString()}`);
  console.log(`   Valid for:  6 months`);
  console.log('');
  console.log('💡 Use this JWT as the "Client Secret" in Supabase Dashboard → Authentication → Providers → Apple');
  console.log('');

} catch (error) {
  console.error('❌ Error generating JWT:', error.message);
  console.error('');
  console.error('Make sure:');
  console.error('  1. The .p8 file is valid');
  console.error('  2. The Team ID is correct');
  console.error('  3. The Service ID is correct');
  process.exit(1);
}
